File: apps/ios/OuraApp/SleepStaging.swift
#if TORCH
import Foundation
// sqlite3 comes from the bridging header (TorchBridge.h includes <sqlite3.h>)

// On-device sleep staging: read the synced DB, assemble the raw SleepNet inputs
// (a faithful port of tools/run_sleep_model.py — the model bakes in its own
// preprocessing), and run sleepnet_moonstone via the LibTorch lite bridge. Returns
// date-label → per-30s stage codes, matching the web dashboard's `nights[].stages`.
enum SleepStaging {
    /// Everything SleepNet receives for one night — also the exact material the
    /// incremental cache fingerprints.
    struct NightInputs {
        let startDs: Int64, endDs: Int64
        let startMs: Int64, endMs: Int64
        let beats: [(Int64, Float, Float, Float)]
        let acm: [(Int64, Float)]
        let temp: [(Int64, Float)]
    }

    // Returns date-key → stage codes, plus a non-nil `error` only for genuine failures
    // (bundled model missing). An empty map with `error == nil` just means no sleep data.
    static func run(nights: [NightRow], events: EventStore.Events, clock: EventStore.RingClock,
                    force: Bool = false, pruneCache: Bool = true,
                    progress: @escaping @Sendable (String) -> Void = { _ in }) -> (staged: [String: [Int]], error: String?) {
        guard let modelPath = Bundle.main.path(forResource: "sleepnet_moonstone_1_2_0", ofType: "ptl")
        else { return ([:], "sleep model file missing from the app bundle") }
        guard !events.isEmpty else { return ([:], nil) }

        // The shared Rust summary canonicalizes brief wake splits and premature
        // bedtime ends. Consume those exact windows so the model cannot reintroduce
        // the raw ring boundary that the UI already corrected.
        // `nights` arrive newest-first from the summary; keep that order so the
        // night the user is looking at stages first.
        let beds = nights.compactMap { night -> (start: Int64, end: Int64, cu: Int64)? in
            guard let start = night.start_ds, let end = night.end_ds else { return nil }
            let captured = events.restricted("tag=118").first { event in
                event.tag == 0x76
                    && (event.json["bedtime_start_ds"] as? NSNumber)?.int64Value == start
            }?.cu
            return (start, end, captured ?? events.last?.cu ?? 0)
        }

        // A night's fingerprint covers its exact model inputs; a hit is by
        // construction the hypnogram the model would have produced.
        func fingerprint(_ inputs: NightInputs) -> String {
            var h = FNV64()
            h.combine(inputs.startDs); h.combine(inputs.endDs)
            h.combine(inputs.startMs); h.combine(inputs.endMs)  // RingClock re-dating backstop
            h.combine(inputs.beats.count)
            for b in inputs.beats { h.combine(b.0); h.combine(b.1); h.combine(b.2); h.combine(b.3) }
            h.combine(inputs.acm.count)
            for a in inputs.acm { h.combine(a.0); h.combine(a.1) }
            h.combine(inputs.temp.count)
            for t in inputs.temp { h.combine(t.0); h.combine(t.1) }
            return h.hex
        }
        let globalKey = ModelCacheStore.globalKey(profile: nil)
        var cache: [String: StagedNightEntry] = ModelCacheStore.load(ModelCacheStore.stagingFile,
                                                                     globalKey: globalKey)

        // key by the exact bedtime start_ds (matches the summary's night.start_ds)
        // so two sleeps on one calendar day stay distinct.
        var result: [String: [Int]] = [:]
        var currentKeys = Set<String>()
        var recomputed = 0
        // Count only the nights that actually need the model, so the progress pill
        // reads "night 1 of 1" on a normal launch instead of the whole history.
        var pending: [(key: String, fp: String, inputs: NightInputs)] = []
        for bed in beds {
            guard !AnalysisRun.cancelled, events.error == nil else { return ([:], "analysis interrupted") }
            guard let inputs = nightInputs(start: bed.start, end: bed.end, cu: bed.cu,
                                           events: events, clock: clock) else { continue }
            guard events.error == nil else { return ([:], "event read failed") }
            let key = String(bed.start), fp = fingerprint(inputs)
            currentKeys.insert(key)
            if !force, let entry = cache[key], entry.fp == fp {
                if !entry.stages.isEmpty { result[key] = entry.stages }
            } else {
                pending.append((key, fp, inputs))
            }
        }
        for (index, night) in pending.enumerated() {
            let key = night.key, fp = night.fp, inputs = night.inputs
            progress(pending.count > 1 ? "Analyzing sleep \(index + 1)/\(pending.count)" : "Analyzing sleep")
            guard !AnalysisRun.cancelled else { return ([:], "analysis paused") }
            guard let stages = stageNight(inputs, modelPath: modelPath) else { return (result, "sleep inference failed") }
            // A manual refresh with insufficient data must not erase a useful
            // cached hypnogram. The caller reports that no result was produced.
            if force && stages.isEmpty { continue }
            result[key] = stages
            cache[key] = StagedNightEntry(fp: fp, stages: stages)
            ModelCacheStore.save(ModelCacheStore.stagingFile, globalKey: globalKey, entries: cache)
            recomputed += 1
        }
        guard events.error == nil, !AnalysisRun.cancelled else { return ([:], "analysis interrupted") }
        dlog("models", "sleep recomputed=\(recomputed) total=\(currentKeys.count)")
        let pruned = cache.filter { currentKeys.contains($0.key) }
        if pruneCache && pruned.count != cache.count {
            ModelCacheStore.save(ModelCacheStore.stagingFile, globalKey: globalKey, entries: pruned)
        }
        return (result, nil)
    }

    /// Gather one night's raw model inputs; nil when there's no usable beat data.
    static func nightInputs(start startDs: Int64, end endDs: Int64, cu bedCu: Int64,
                            events: EventStore.Events, clock: EventStore.RingClock) -> NightInputs? {
        // ms(ds) → absolute epoch ms, epoch-aware (ds resets on ring reboot; see EventStore)
        func ms(_ ds: Int64, _ cu: Int64) -> Int64 {
            Int64(clock.unixSeconds(ds, capturedUnix: cu) * 1000)
        }
        let lo = startDs - 6000, hi = endDs + 6000
        var beats: [(Int64, Float, Float, Float)] = []
        var acm: [(Int64, Float)] = [], temp: [(Int64, Float)] = []
        for e in events.restricted("ring_timestamp BETWEEN \(lo) AND \(hi) AND tag IN (96,128,71,70)") {
            switch e.tag {
            case 0x60, 0x80:
                guard let ibi = e.json["ibi_ms"] as? [NSNumber] else { continue }
                let amp = (e.json["amplitude"] as? [NSNumber]) ?? []
                let t = ms(e.ds, e.cu); var acc: Int64 = 0
                for (i, xn) in ibi.enumerated() {
                    let x = xn.int64Value
                    if x <= 0 { continue }
                    acc += x
                    let valid: Float = (x >= 300 && x <= 2000) ? 1 : 0
                    beats.append((t + acc, Float(x), i < amp.count ? amp[i].floatValue : 0, valid))
                }
            case 0x47:
                if let mo = (e.json["motion_seconds"] as? NSNumber)?.floatValue { acm.append((ms(e.ds, e.cu), mo)) }
            case 0x46:
                if let temps = e.json["temps_c"] as? [NSNumber], let c = temps.first?.floatValue { temp.append((ms(e.ds, e.cu), c)) }
            default: break
            }
        }
        beats.sort { $0.0 < $1.0 }; acm.sort { $0.0 < $1.0 }; temp.sort { $0.0 < $1.0 }
        guard !beats.isEmpty, beats.contains(where: { $0.3 == 1 }) else { return nil }
        return NightInputs(startDs: startDs, endDs: endDs,
                           startMs: ms(startDs, bedCu), endMs: ms(endDs, bedCu),
                           beats: beats, acm: acm, temp: temp)
    }

    /// One SleepNet inference over one night's inputs.
    private static func stageNight(_ inputs: NightInputs, modelPath: String) -> [Int]? {
        var ibiTs = inputs.beats.map { $0.0 }
        var ibiVal = inputs.beats.flatMap { [$0.1, $0.2, $0.3] }
        var acmTs = inputs.acm.map { $0.0 }, acmVal = inputs.acm.map { $0.1 }
        var tempTs = inputs.temp.map { $0.0 }, tempVal = inputs.temp.map { $0.1 }
        var out = [Int32](repeating: 0, count: 8192)
        let n = oura_sleepnet(modelPath,
                              &ibiTs, &ibiVal, Int32(inputs.beats.count),
                              &acmTs, &acmVal, Int32(inputs.acm.count),
                              &tempTs, &tempVal, Int32(inputs.temp.count),
                              inputs.startMs, inputs.endMs, &out, 8192)
        guard n >= 0 else { return nil }
        return out.prefix(Int(n)).map(Int.init)
    }
}
#endif

File: apps/android/app/src/full/java/md/thomas/openoura/models/SleepStaging.kt
package md.thomas.openoura.models

import android.content.Context
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonPrimitive
import md.thomas.openoura.data.EventStore
import md.thomas.openoura.data.NightRow
import md.thomas.openoura.diag.Diagnostics.log

/**
 * On-device sleep staging: read the synced DB, assemble the raw SleepNet inputs (a
 * faithful port of tools/run_sleep_model.py — the model bakes in its own preprocessing),
 * and run sleepnet_moonstone through the LibTorch lite bridge. Returns
 * bedtime-start_ds → per-30s stage codes, matching `nights[].stages` on the other clients.
 *
 * Port of apps/ios/OuraApp/SleepStaging.swift.
 */
object SleepStaging {

    private const val MODEL = "sleepnet_moonstone_1_2_0"

    /** Everything SleepNet receives for one night — also the material the cache fingerprints. */
    class NightInputs(
        val startDs: Long,
        val endDs: Long,
        val startMs: Long,
        val endMs: Long,
        /** (timestampMs, ibiMs, amplitude, valid) */
        val beats: List<Beat>,
        val acm: List<Sample>,
        val temp: List<Sample>,
    )

    class Beat(val t: Long, val ibi: Float, val amplitude: Float, val valid: Float)
    class Sample(val t: Long, val v: Float)

    class Result(val staged: Map<String, List<Int>>, val error: String?)

    /**
     * Returns start_ds → stage codes, plus a non-null [Result.error] only for genuine
     * failures (missing model asset). An empty map with a null error just means there is
     * no sleep data to stage.
     */
    fun run(
        context: Context,
        nights: List<NightRow>,
        events: List<EventStore.Ev>,
        clock: EventStore.RingClock,
        progress: (String) -> Unit = {},
    ): Result {
        val modelPath = ModelAssets.path(context, MODEL)
            ?: return Result(emptyMap(), "sleep model file missing from the app assets")
        if (events.isEmpty()) return Result(emptyMap(), null)

        // The shared Rust summary canonicalizes brief wake splits and premature bedtime
        // ends. Consume those exact windows so the model cannot reintroduce the raw ring
        // boundary the UI already corrected. `nights` arrive newest-first from the
        // summary; keep that order so the night the user is looking at stages first.
        val beds = nights.mapNotNull { night ->
            val start = night.startDs ?: return@mapNotNull null
            val end = night.endDs ?: return@mapNotNull null
            val captured = events.firstOrNull { e ->
                e.tag == 0x76 && e.long("bedtime_start_ds") == start
            }?.cu ?: events.lastOrNull()?.cu ?: 0L
            Triple(start, end, captured)
        }

        val globalKey = ModelCacheStore.globalKey(context, null)
        val cache = ModelCacheStore.load(
            context, ModelCacheStore.STAGING_FILE, globalKey, StagedNightEntry.serializer()
        )

        val result = HashMap<String, List<Int>>()
        val dirty = ArrayList<Triple<String, NightInputs, String>>()
        val currentKeys = HashSet<String>()

        for ((start, end, cu) in beds) {
            val inputs = nightInputs(start, end, cu, events, clock) ?: continue
            // Key by the exact bedtime start_ds (matching the summary's night.start_ds)
            // so two sleeps on one calendar day stay distinct.
            val key = start.toString()
            val fp = fingerprint(inputs)
            currentKeys.add(key)
            val entry = cache[key]
            if (entry != null && entry.fp == fp) {
                if (entry.stages.isNotEmpty()) result[key] = entry.stages
            } else {
                dirty.add(Triple(key, inputs, fp))
            }
        }

        log("models", "staging: ${dirty.size}/${currentKeys.size} nights to recompute")
        dirty.forEachIndexed { index, (key, inputs, fp) ->
            progress("staging sleep · night ${index + 1}/${dirty.size}")
            // Empty stages are cached too: a night the model can't stage shouldn't be
            // retried on every reload until its inputs change.
            val stages = stageNight(inputs, modelPath) ?: emptyList()
            if (stages.isNotEmpty()) result[key] = stages
            cache[key] = StagedNightEntry(fp, stages)
            ModelCacheStore.save(
                context, ModelCacheStore.STAGING_FILE, globalKey, cache,
                StagedNightEntry.serializer(),
            )
        }

        val pruned = cache.filterKeys { it in currentKeys }
        if (pruned.size != cache.size) {
            ModelCacheStore.save(
                context, ModelCacheStore.STAGING_FILE, globalKey, pruned,
                StagedNightEntry.serializer(),
            )
        }
        return Result(result, null)
    }

    /** Gather one night's raw model inputs; null when there is no usable beat data. */
    fun nightInputs(
        startDs: Long,
        endDs: Long,
        bedCu: Long,
        events: List<EventStore.Ev>,
        clock: EventStore.RingClock,
    ): NightInputs? {
        // ms(ds) → absolute epoch ms, epoch-aware (ds resets on ring reboot; see EventStore)
        fun ms(ds: Long, cu: Long): Long = (clock.unixSeconds(ds, cu) * 1000).toLong()

        val lo = startDs - 6000
        val hi = endDs + 6000
        val beats = ArrayList<Beat>()
        val acm = ArrayList<Sample>()
        val temp = ArrayList<Sample>()

        for (e in events) {
            if (e.ds < lo || e.ds > hi) continue
            when (e.tag) {
                0x60, 0x80 -> {
                    val ibi = (e.json?.get("ibi_ms") as? JsonArray) ?: continue
                    val amp = e.json["amplitude"] as? JsonArray
                    val t = ms(e.ds, e.cu)
                    var acc = 0L
                    ibi.forEachIndexed { i, element ->
                        val x = element.jsonPrimitive.doubleOrNull?.toLong() ?: 0L
                        if (x > 0) {
                            acc += x
                            val valid = if (x in 300..2000) 1f else 0f
                            val amplitude = amp?.getOrNull(i)
                                ?.jsonPrimitive?.doubleOrNull?.toFloat() ?: 0f
                            beats.add(Beat(t + acc, x.toFloat(), amplitude, valid))
                        }
                    }
                }
                0x47 -> e.num("motion_seconds")?.let { acm.add(Sample(ms(e.ds, e.cu), it.toFloat())) }
                0x46 -> {
                    val temps = e.json?.get("temps_c") as? JsonArray
                    temps?.firstOrNull()?.jsonPrimitive?.doubleOrNull?.let {
                        temp.add(Sample(ms(e.ds, e.cu), it.toFloat()))
                    }
                }
            }
        }

        beats.sortBy { it.t }
        acm.sortBy { it.t }
        temp.sortBy { it.t }
        if (beats.isEmpty() || beats.none { it.valid == 1f }) return null

        return NightInputs(
            startDs = startDs,
            endDs = endDs,
            startMs = ms(startDs, bedCu),
            endMs = ms(endDs, bedCu),
            beats = beats,
            acm = acm,
            temp = temp,
        )
    }

    /**
     * A night's fingerprint covers its exact model inputs; a hit is by construction the
     * hypnogram the model would have produced.
     */
    private fun fingerprint(inputs: NightInputs): String {
        val h = FNV64()
        h.combine(inputs.startDs)
        h.combine(inputs.endDs)
        h.combine(inputs.startMs)  // RingClock re-dating backstop
        h.combine(inputs.endMs)
        h.combine(inputs.beats.size)
        for (b in inputs.beats) {
            h.combine(b.t); h.combine(b.ibi); h.combine(b.amplitude); h.combine(b.valid)
        }
        h.combine(inputs.acm.size)
        for (a in inputs.acm) { h.combine(a.t); h.combine(a.v) }
        h.combine(inputs.temp.size)
        for (t in inputs.temp) { h.combine(t.t); h.combine(t.v) }
        return h.hex
    }

    /** One SleepNet inference over one night's inputs. */
    private fun stageNight(inputs: NightInputs, modelPath: String): List<Int>? {
        val ibiTs = LongArray(inputs.beats.size) { inputs.beats[it].t }
        val ibiVal = FloatArray(inputs.beats.size * 3)
        inputs.beats.forEachIndexed { i, b ->
            ibiVal[i * 3] = b.ibi
            ibiVal[i * 3 + 1] = b.amplitude
            ibiVal[i * 3 + 2] = b.valid
        }
        val acmTs = LongArray(inputs.acm.size) { inputs.acm[it].t }
        val acmVal = FloatArray(inputs.acm.size) { inputs.acm[it].v }
        val tempTs = LongArray(inputs.temp.size) { inputs.temp[it].t }
        val tempVal = FloatArray(inputs.temp.size) { inputs.temp[it].v }

        val out = TorchBridge.sleepnet(
            modelPath,
            ibiTs, ibiVal,
            acmTs, acmVal,
            tempTs, tempVal,
            inputs.startMs, inputs.endMs,
        ) ?: return null
        return if (out.isEmpty()) null else out.toList()
    }
}
